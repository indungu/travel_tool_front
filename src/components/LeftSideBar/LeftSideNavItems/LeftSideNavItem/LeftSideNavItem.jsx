import React, { Fragment } from 'react';
import { NavLink, withRouter } from 'react-router-dom';
import { PropTypes } from 'prop-types';
import ActiveDropdrown from '../../../../images/icons/dropdown_active.svg';
import InactiveDropdown from '../../../../images/icons/dropdown_inactive.svg';
import './LeftSideNavItem.scss';

export class LeftSideNavItem extends React.PureComponent {
  static propTypes = {
    text: PropTypes.string.isRequired,
    link_to: PropTypes.string.isRequired,
    linkIcons: PropTypes.object.isRequired,
    isDropdown: PropTypes.bool,
    children: PropTypes.arrayOf(PropTypes.object.isRequired),
    match: PropTypes.object.isRequired
  }

  static defaultProps = {
    children: [{}],
    isDropdown: false
  }

  constructor(props) {
    super(props);
    this.state = {
      dropdownActive: false,
    };
  }

  componentWillReceiveProps(nextProps, nextContext) {
    if(nextContext.activeNavItem !== this)
      this.setState({dropdownActive: false});
  }

  getDropdownElements = (children, dropdownStatus, linkIsActive) => {
    const dropdownItems = (
      <div className={`LeftSideNavItem__dropdownContent ${dropdownStatus}`}>
        {children}
      </div>
    );

    const dropdownIcon = (
      <div
        className={`dropdownIcon dropdownIcon--${dropdownStatus}`}
        style={{position: 'absolute', right: '30px', top: '17px'}}>
        <img src={linkIsActive? ActiveDropdrown: InactiveDropdown} alt="dropdown" />
      </div>
    );

    return {
      items: dropdownItems,
      icon: dropdownIcon
    };
  }

  isActive = (activeNavItem, link_to) => {
    const {match} = this.props;
    return (activeNavItem === this) && match.path.startsWith(link_to);
  }

  toggleDropdown = () => {
    const {setActiveNavItem} = this.context;
    setActiveNavItem(this);
    const {dropdownActive} = this.state;
    this.setState({dropdownActive: !dropdownActive});
  }

  render() {
    const {text, link_to, linkIcons, isDropdown, children, match } = this.props;
    const {activeNavItem} = this.context;
    const linkIsActive = this.isActive(activeNavItem, link_to);
    const { dropdownActive } = this.state;
    let dropdownStatus = dropdownActive? 'active': 'inactive';
    let dropdownElements;

    if (isDropdown)
      dropdownElements = this.getDropdownElements(children, dropdownStatus, linkIsActive);

    return (
      <Fragment>
        <NavLink onClick={this.toggleDropdown} to={link_to} className="LeftSideNavItem">
          <div className="LeftSideNavItem__leftIcon">
            <img src={linkIsActive? linkIcons.active: linkIcons.inactive} alt="icon" />
          </div>
          <div className="LeftSideNavItem__Nav-text">
            {text}
          </div>
          { dropdownElements? dropdownElements.icon: null}
        </NavLink>
        {dropdownElements? dropdownElements.items: null}
      </Fragment>
    );
  }
}

LeftSideNavItem.contextTypes = {
  activeNavItem: PropTypes.object.isRequired,
  setActiveNavItem: PropTypes.func.isRequired
}
export default withRouter(LeftSideNavItem);
