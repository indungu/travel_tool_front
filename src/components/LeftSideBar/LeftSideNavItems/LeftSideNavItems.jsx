import React, { PureComponent, Fragment } from 'react';
import { PropTypes } from 'prop-types';
import LeftSidebarNavItem from './LeftSideNavItem/LeftSideNavItem';
import DropdownItem from './DropdownItem/DropdownItem';


class LeftSideNavItems extends PureComponent {

  getChildContext() {
    const {setActiveNavItem, activeNavItem} = this.props;
    return {
      activeNavItem,
      setActiveNavItem
    }
  }

  render() {
    const { navIconSources, activeNavItem, setActiveNavItem} = this.props;
    return (
      <Fragment>
        <LeftSidebarNavItem isDropdown linkIcons={navIconSources.requestsIcon} link_to="/requests" text="Requests">
          <DropdownItem link_to="/requests/my-requests">
            My Requests
          </DropdownItem>
          <DropdownItem link_to="/requests/my-approvals">
            My Approvals
          </DropdownItem>
        </LeftSidebarNavItem>
        <LeftSidebarNavItem linkIcons={navIconSources.settingsIcon} link_to="/settings" text="Settings" />
      </Fragment>
    );
  }
}

LeftSideNavItems.childContextTypes = {
  activeNavItem: PropTypes.object.isRequired,
  setActiveNavItem: PropTypes.func.isRequired
}

export default LeftSideNavItems;
