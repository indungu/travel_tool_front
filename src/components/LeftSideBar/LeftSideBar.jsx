import React, {Component} from 'react';
import LeftSideNavItems from './LeftSideNavItems/LeftSideNavItems';
import ActiveBookmarkIcon from '../../images/icons/bookmark_active.svg';
import InactiveBookmarkIcon from '../../images/icons/bookmark_inactive.svg';
import ActiveSettingsIcon from '../../images/icons/settings_active.svg';
import InactiveSettingsIcon from '../../images/icons/settings_inactive.svg';
import './LeftSideBar.scss';

class LeftSideBar extends Component {

  constructor(props) {
    super(props);
    this.state = {
      activeNavItem: {}
    };
  }

  setActiveNavItem = (clickedNavItem) => {
    this.setState({activeNavItem: clickedNavItem});
  }

  getNavigationIcons() {
      return {
        requestsIcon: {
          active: ActiveBookmarkIcon,
          inactive: InactiveBookmarkIcon
        },
        settingsIcon: {
          active: ActiveSettingsIcon,
          inactive: InactiveSettingsIcon
        },
      }
  }

  render() {
    const navIconSources = this.getNavigationIcons();
    const {activeNavItem} = this.state;

    return (
      <div className="LeftSideBar mdl-cell mdl-cell--2-col">
        <div className="LeftSideBar__fixed_wrapper">
          <div style={{overflowY: 'scroll', height: '100%'}}>
            <LeftSideNavItems
              activeNavItem={activeNavItem}
              navIconSources={navIconSources}
              setActiveNavItem={this.setActiveNavItem} />
          </div>
        </div>
      </div>
    );
  }
}

export default LeftSideBar;
