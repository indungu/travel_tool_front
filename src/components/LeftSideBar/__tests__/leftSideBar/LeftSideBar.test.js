import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import LeftSideBar from '../../LeftSideBar';
import LeftSideNavItems from '../../LeftSideNavItems/LeftSideNavItems';
import { LeftSideNavItem } from '../../LeftSideNavItems/LeftSideNavItem/LeftSideNavItem';
import DropdownItem from '../../LeftSideNavItems/DropdownItem/DropdownItem';


describe('<LeftSideBar />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<LeftSideBar />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('<LeftSideNavItems />', () => {
    let wrapper;

    beforeEach(() => {
      const navIconSources = {
        requestsIcon: {
          active: 'test.svg',
          inactive: 'test.svg'
        },
        settingsIcon: {
          active: 'test.svg',
          inactive: 'test.svg'
        },
      };
      wrapper = shallow(
        <LeftSideNavItems
          activeNavItem={{}}
          navIconSources={navIconSources}
          setActiveNavItem={jest.fn}
        />
      );
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('it renders two dropdown items', () => {
      const dropDownItems = wrapper.find('DropdownItem');
      expect(dropDownItems).toHaveLength(2);
    });
  });
});

describe('<LeftSideNavItem />', () => {

  let match, iconSrc, wrapper;

  beforeEach(() => {
    match = {path: '/test-link'};
    iconSrc = () => {};
    const context = {
      setActiveNavItem: jest.fn,
      activeNavItem: {}
    };
    wrapper = shallow(
      <LeftSideNavItem
        text="Test link"
        isDropdown
        linkIcons={{active: 'test.svg', inactive: 'test.svg'}}
        match={match}
        link_to="/test-link" />, {context}
    );
  });

  it('renders the LeftSideNavItem', () => {
    expect(wrapper).toMatchSnapshot();
  });


  describe('<LeftSideNavItem /> with dropdown', () => {
    beforeEach(() => {
      wrapper.setProps({isDropdown: true});
    });

    it('renders as dropdown', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('toggles dropdown', () => {
      const expectedStateOnClick = { dropdownActive: true };

      wrapper.find('NavLink').simulate('click');
      const stateOnClick = wrapper
        .instance()
        .state;
      expect(stateOnClick).toEqual(expectedStateOnClick);
    });

    it('closes dropdown its when it is not the active item', () => {
      const expectedStateOnClick = { dropdownActive: false };
      wrapper.setContext({activeNavItem: {}});
      const stateOnNewProps = wrapper.instance().state;
      expect(stateOnNewProps).toEqual(expectedStateOnClick);
    });
  });
});

describe('<DropdownItem />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <DropdownItem link_to="/test-link/test-endpoint">
        Test Link
      </DropdownItem>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
