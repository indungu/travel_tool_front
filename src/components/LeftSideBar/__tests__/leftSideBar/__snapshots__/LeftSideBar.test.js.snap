// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownItem /> renders correctly 1`] = `
<NavLink
  activeClassName="active"
  aria-current="page"
  className="DropdownItem"
  to="/test-link/test-endpoint"
>
  Test Link
</NavLink>
`;

exports[`<LeftSideBar /> <LeftSideNavItems /> renders correctly 1`] = `
<React.Fragment>
  <withRouter(LeftSideNavItem)
    isDropdown={true}
    linkIcons={
      Object {
        "active": "test.svg",
        "inactive": "test.svg",
      }
    }
    link_to="/requests"
    text="Requests"
  >
    <DropdownItem
      link_to="/requests/my-requests"
    >
      My Requests
    </DropdownItem>
    <DropdownItem
      link_to="/requests/my-approvals"
    >
      My Approvals
    </DropdownItem>
  </withRouter(LeftSideNavItem)>
  <withRouter(LeftSideNavItem)
    linkIcons={
      Object {
        "active": "test.svg",
        "inactive": "test.svg",
      }
    }
    link_to="/settings"
    text="Settings"
  />
</React.Fragment>
`;

exports[`<LeftSideBar /> renders correctly 1`] = `
<div
  className="LeftSideBar mdl-cell mdl-cell--2-col"
>
  <div
    className="LeftSideBar__fixed_wrapper"
  >
    <div
      style={
        Object {
          "height": "100%",
          "overflowY": "scroll",
        }
      }
    >
      <LeftSideNavItems
        activeNavItem={Object {}}
        navIconSources={
          Object {
            "requestsIcon": Object {
              "active": "bookmark_active.svg",
              "inactive": "bookmark_inactive.svg",
            },
            "settingsIcon": Object {
              "active": "settings_active.svg",
              "inactive": "settings_inactive.svg",
            },
          }
        }
        setActiveNavItem={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`<LeftSideNavItem /> <LeftSideNavItem /> with dropdown renders as dropdown 1`] = `
<React.Fragment>
  <NavLink
    activeClassName="active"
    aria-current="page"
    className="LeftSideNavItem"
    onClick={[Function]}
    to="/test-link"
  >
    <div
      className="LeftSideNavItem__leftIcon"
    >
      <img
        alt="icon"
        src="test.svg"
      />
    </div>
    <div
      className="LeftSideNavItem__Nav-text"
    >
      Test link
    </div>
    <div
      className="dropdownIcon dropdownIcon--inactive"
      style={
        Object {
          "position": "absolute",
          "right": "30px",
          "top": "17px",
        }
      }
    >
      <img
        alt="dropdown"
        src="dropdown_inactive.svg"
      />
    </div>
  </NavLink>
  <div
    className="LeftSideNavItem__dropdownContent inactive"
  />
</React.Fragment>
`;

exports[`<LeftSideNavItem /> renders the LeftSideNavItem 1`] = `
<React.Fragment>
  <NavLink
    activeClassName="active"
    aria-current="page"
    className="LeftSideNavItem"
    onClick={[Function]}
    to="/test-link"
  >
    <div
      className="LeftSideNavItem__leftIcon"
    >
      <img
        alt="icon"
        src="test.svg"
      />
    </div>
    <div
      className="LeftSideNavItem__Nav-text"
    >
      Test link
    </div>
    <div
      className="dropdownIcon dropdownIcon--inactive"
      style={
        Object {
          "position": "absolute",
          "right": "30px",
          "top": "17px",
        }
      }
    >
      <img
        alt="dropdown"
        src="dropdown_inactive.svg"
      />
    </div>
  </NavLink>
  <div
    className="LeftSideNavItem__dropdownContent inactive"
  />
</React.Fragment>
`;
